### ISSUE TITLE
Bots to simplify project managment

### ISSUE THREAD

**Issue by squakez** [apache/camel-k#2623]
Hi folks,
I was having a look at the dependencies that we manage in the project and I saw we're quite out of sync with the most recent version upgrades for almost all dependencies:

![camel-k-deps](https://user-images.githubusercontent.com/23169723/132493469-93d4b1fe-5fdd-48a5-a571-a1387acc0f40.png)

Although it's not the perfect tool and may be not valid for certain dependencies that we want to stick to certain versions (ie, knative or kubernetes), I was wondering if we should start introducing some bot to manage dependencies, as [dependabot](https://dependabot.com/). I think it will help us, maybe not immediately as we will have may PRs to review, but in the long run we should have good results.

Also, another bot I'd like to introduce is a [stale bot](https://github.com/apps/stale) in order to automatically close inactive issues. I'm finding quite difficult to browse along issues that seems are no longer needed/meaningful.

Thoughts?

**Comment by oscerd** [comment]
I think it makes sense to use dependabot in go projects. We need to be sure to avoid update on fundamental dependencies, as you said, so +1.

**Comment by astefanutti** [comment]
> Although it's not the perfect tool and may be not valid for certain dependencies that we want to stick to certain versions (ie, knative or kubernetes)

Most of the dependencies are transitively inherited from the Kubernetes version. Actually, the tree is mostly k8s / client-go > controller-runtime | Knative. As you've nailed it, we want to control these. Actually I don't think dependabot would be able to resolve the conflicts between Knative / controller-runtime, and our own schedule of k8s versions support. Not to mention k8s support by KinD for testing. So I anticipate more noise from dependabot than added-value. We can try to prove the above wrong, otherwise that'd be -1 for me.

On the other-hand, +1 for stale bot 😃.

**Comment by astefanutti** [comment]
For example, right now, we cannot upgrade to controller-runtime 0.9.0, because it depends on k8s 1.21, while the latest Knative version still depends on k8s 1.20. And client-go 1.21 introduces a breaking change. So we have to wait until Knative release a new version (the upgrade has been done in `main` branch recently).

From my experience, this situation happens quite often, almost at every k8s upgrade, and I don't think dependabot is capable of handling these. For instance, in your example, it suggests to upgrade to k8s 1.22.

**Comment by lburgazzoli** [comment]
> > Although it's not the perfect tool and may be not valid for certain dependencies that we want to stick to certain versions (ie, knative or kubernetes)
> 
> Most of the dependencies are transitively inherited from the Kubernetes version. Actually, the tree is mostly k8s / client-go > controller-runtime | Knative. As you've nailed it, we want to control these. Actually I don't think dependabot would be able to resolve the conflicts between Knative / controller-runtime, and our own schedule of k8s versions support. Not to mention k8s support by KinD for testing. So I anticipate more noise from dependabot than added-value. We can try to prove the above wrong, otherwise that'd be -1 for me.
> 

Not sure how feasible and how much value it will bring, but eventually those "problematic" deps can be excluded from dependabot checks

**Comment by astefanutti** [comment]
> > > Although it's not the perfect tool and may be not valid for certain dependencies that we want to stick to certain versions (ie, knative or kubernetes)
> > 
> > 
> > Most of the dependencies are transitively inherited from the Kubernetes version. Actually, the tree is mostly k8s / client-go > controller-runtime | Knative. As you've nailed it, we want to control these. Actually I don't think dependabot would be able to resolve the conflicts between Knative / controller-runtime, and our own schedule of k8s versions support. Not to mention k8s support by KinD for testing. So I anticipate more noise from dependabot than added-value. We can try to prove the above wrong, otherwise that'd be -1 for me.
> 
> Not sure how feasible and how much value it will bring, but eventually those "problematic" deps can be excluded from dependabot checks

I fear it would be equivalent to excluding 80% of the project dependencies, which would defeat the purpose of it.

**Comment by astefanutti** [comment]
I'd also argue that dependency management is too critical for a project like Camel K, that evolves in a fast-moving ecosystem, to be delegated to tools as naive as dependabot. You also miss the opportunity to learn new APIs and concepts introduced in newer versions. In that fast-moving context, a _qualitative_ upgrade is seldom just a version bump. I'd rather keep Camel K at the cutting edge, granted it requires some worthy manual efforts from time to time, than considering Camel K already ready to ride the slope of the "plateau of productivity" 😃.

**Comment by squakez** [comment]
Thanks for the comments. I see the arguments and I guess we can keep the dependabot out of scope for the time being. However, I see no objections on the stale bot. If that is okey with the rest of people I'll proceed in the next days to enable it.

**Comment by apupier** [comment]
> You also miss the opportunity to learn new APIs and concepts introduced in newer versions.

The PRs are created with links to changelog and new commits

> I fear it would be equivalent to excluding 80% of the project dependencies, which would defeat the purpose of it.

it is already a win for 20% of the dependencies and does not impact the 80 other percent.

> I'd also argue that dependency management is too critical for a project like Camel K, that evolves in a fast-moving ecosystem, to be delegated to tools as naive as dependabot. 

I tend to think the exact opposite. Because it is a fast-moving ecosystem, we need to try to follow-up on latest versions and detect incompatibilities even faster than slow-paced projects.

**Comment by apupier** [comment]
> For example, right now, we cannot upgrade to controller-runtime 0.9.0, because it depends on k8s 1.21, while the latest Knative version still depends on k8s 1.20. And client-go 1.21 introduces a breaking change. So we have to wait until Knative release a new version (the upgrade has been done in `main` branch recently).
> 
> From my experience, this situation happens quite often, almost at every k8s upgrade, and I don't think dependabot is capable of handling this. For instance, in your example, it suggests to upgrade to k8s 1.22 and controller-runtime 0.9.0, which is not possible ATM.

vote for it: https://github.com/dependabot/dependabot-core/issues/1296 (even k8s case specifically mentioned https://github.com/dependabot/dependabot-core/issues/1296#issuecomment-902195531 )

**Comment by astefanutti** [comment]
> > You also miss the opportunity to learn new APIs and concepts introduced in newer versions.
> 
> The PRs are created with links to changelog and new commits
> 
> > I fear it would be equivalent to excluding 80% of the project dependencies, which would defeat the purpose of it.
> 
> it is already a win for 20% of the dependencies and does not impact the 80 other percent.
> 
> > I'd also argue that dependency management is too critical for a project like Camel K, that evolves in a fast-moving ecosystem, to be delegated to tools as naive as dependabot.
> 
> I tend to think the exact opposite. Because it is a fast-moving ecosystem, we need to try to follow-up on latest versions and detect incompatibilities even faster than slow-paced projects.

This is exactly what we are doing. The issue statement forget to mention that very point! So it's an issue for a problem that does not really exists, and propose to apply a one-size fits all solution that is less efficient that the current process. For instance, the screenshot of the proposed upgrades is totally inapplicable, because the logic is so naive. On the other hand we did the upgrade with #2655 in less that 24 hours after it became possible.

**Comment by apupier** [comment]
I randomly picked some. These dependencies cannot be upgraded or they were missed? I find strange that all of them cannot be upgraded.
* [github.com/go-logr/logr v0.4.0 ](https://github.com/apache/camel-k/blob/2aea4c7ee65fd48c51a783c9de2ff4978b992306/go.mod#L16) new release available since 7 july
* [github.com/Masterminds/semver v1.5.0](https://github.com/apache/camel-k/blob/2aea4c7ee65fd48c51a783c9de2ff4978b992306/go.mod#L6)  new release available since 12 September 2019
*  [github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring 0.42.1 ](https://github.com/apache/camel-k/blob/2aea4c7ee65fd48c51a783c9de2ff4978b992306/go.mod#L28) new release available since 23 September 2020
* [go.uber.org/multierr v1.6.0](https://github.com/apache/camel-k/blob/2aea4c7ee65fd48c51a783c9de2ff4978b992306/go.mod#L44) new release available since 6 May
* [github.com/spf13/cobra v1.1.3](https://github.com/apache/camel-k/blob/2aea4c7ee65fd48c51a783c9de2ff4978b992306/go.mod#L39)  new release available since 1st July
* [github.com/operator-framework/api v0.3.8](https://github.com/apache/camel-k/blob/2aea4c7ee65fd48c51a783c9de2ff4978b992306/go.mod#L26) new release available since 27 July 2020

**Comment by astefanutti** [comment]
Thanks, these are the 20% :D If it's possible to configure an _allow_ list for these dependencies, so that all the noise generated for the 80% is avoided, we'll be happy to access a PR.

**Comment by squakez** [comment]
> > > You also miss the opportunity to learn new APIs and concepts introduced in newer versions.
> > 
> > 
> > The PRs are created with links to changelog and new commits
> > > I fear it would be equivalent to excluding 80% of the project dependencies, which would defeat the purpose of it.
> > 
> > 
> > it is already a win for 20% of the dependencies and does not impact the 80 other percent.
> > > I'd also argue that dependency management is too critical for a project like Camel K, that evolves in a fast-moving ecosystem, to be delegated to tools as naive as dependabot.
> > 
> > 
> > I tend to think the exact opposite. Because it is a fast-moving ecosystem, we need to try to follow-up on latest versions and detect incompatibilities even faster than slow-paced projects.
> 
> This is exactly what we are doing. The issue statement forgets to mention that very point! So it's an issue for a problem that does not really exists, and propose to apply a one-size fits all solution that is less efficient than the current process. For instance, the screenshot of the proposed upgrades is totally inapplicable, because the logic is so naive. On the other hand we did the upgrade with #2655 in less than 24 hours after it became possible.

Just a note about the screenshot. It was an example taken from VSCode to illustrate the issue :) . It has nothing to do with dependabot and the suggested changes may not be what the bot would pick up.

**Comment by astefanutti** [comment]
> Just a note about the screenshot. It was an example taken from VSCode to illustrate the issue :) . It has nothing to do with dependabot and the suggested changes may not be what the bot would pick up.

Right, I think the issue also fails to identify the problem at hand, which is reducing the latency there could be between the time an upgrade is possible and it's actually done. From that standpoint, I think the project has been doing well as with #2655 , granted it's for the 80% most important dependencies. And for these, I still fail to see how dependabot would help.
